sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala

```diff
@@ -170,4 +171,34 @@ class CollapseProjectSuite extends PlanTest {
     val expected = Sample(0.0, 0.6, false, 11L, relation.select('a as 'c)).analyze
     comparePlans(optimized, expected)
   }
+
+  test("SPARK-32945: avoid collapsing projects if reaching max allowed common exprs") {
```

Comment:
Could you add end-2-end tests having physical planning? I'm worried that this might have the same issue I described in https://github.com/apache/spark/pull/29094#discussion_r458201704

Reply:
Oh, good point. We still change projections even in physical plan...

Reply:
I added similar check into `ScanOperation` rule. I will look at adding tests for physical plan later.

Reply:
Added end-to-end test.